docs/design/taskmanager.md

```diff
@@ -0,0 +1,96 @@
+# Task Manager Design Document
+
+## System Overview
+
+The Task Manager is a robust system for handling asynchronous task execution with state persistence, retry capabilities, and graceful shutdown handling. It's currently integrated with snapshot management functionality.
+
+## Core Components
+
+### 1. Task States
+
+```python
+class TaskState(StrEnum):
+    PENDING = "Pending"
+    RUNNING = "Running"
+    FAILED = "Failed"
+    COMPLETED = "Completed"
+```
+
+### 2. Task Information Structure
+
+```python
+class TaskInfo(TypedDict):
+    task: str          # Task name/identifier
+    args: list         # Positional arguments
+    kwargs: dict       # Keyword arguments
+    retry_count: int   # Number of retry attempts
+    state: TaskState   # Current state of task
+```
+
+### 3. Key Features
+
+#### State Management
+
+- **Persistent Storage**
+  - Tasks stored in JSON format at `{DATA_DIR}/tasks.json`
+  - Thread-safe operations using Lock mechanism
+  - Automatic state recovery after crashes
+  - Handles stuck tasks by marking them as failed
+
+#### Task Execution
+
+- Asynchronous execution using provided executor
+- Task identification using MD5 hash of task parameters
+- Current supported tasks:
+
+  ```python
+  class TaskName(StrEnum):
+      CREATE_SNAPSHOT = "CreateSnapshot"
+  ```
+
+#### Retry Mechanism
+
+- Configurable retry settings:
+  - Retry interval (default: 5 seconds)
+  - Maximum retries (default: 5 attempts)
+- Automatic retry for failed tasks
+- Cleanup of non-retriable and completed tasks
+
+## Technical Implementation
+
+### 1. Thread Safety
+
+- Uses `threading.Lock` for file operations
+- Ensures atomic operations on task storage
+
+### 2. Task Lifecycle
+
+1. Task submission
```

Comment:
Should we also add a max number of tasks? IE with a max of 16 tasks, if we get another create_snapshot or create_clone then we'd reject it

Reply:
We have a limited thread pool, I will do a quick test to see what will happen, they will get queued I think